File: linkmint-backend/eventbus-go/publisher.go
package eventbus

import (
	"context"
	"fmt"
	"log/slog"

	"github.com/twmb/franz-go/pkg/kgo"
	"go.opentelemetry.io/otel/trace"
)

type ctxKey int

const correlationKey ctxKey = 0

// WithCorrelationID returns a context carrying the request correlation id; Publish stamps it into
// the envelope's correlation_id so a trace survives the async hop.
func WithCorrelationID(ctx context.Context, id string) context.Context {
	return context.WithValue(ctx, correlationKey, id)
}

func correlationFrom(ctx context.Context) string {
	if v, ok := ctx.Value(correlationKey).(string); ok {
		return v
	}
	return ""
}

// Publisher produces domain events to Kafka. Its Publish method matches the services' existing
// domain.Publisher seam — Publish(ctx, name, key string, payload any) error — so it drops in for the
// LogPublisher with no call-site change.
type Publisher struct {
	client *kgo.Client
	source string
	log    *slog.Logger
}

// NewPublisher builds a synchronous, idempotent producer (acks=all). source is the producing service
// name, stamped into every envelope.
func NewPublisher(cfg Config, source string, log *slog.Logger) (*Publisher, error) {
	if log == nil {
		log = slog.Default()
	}
	if len(cfg.Brokers) == 0 {
		return nil, fmt.Errorf("eventbus: no kafka brokers configured")
	}
	cl, err := kgo.NewClient(
		kgo.SeedBrokers(cfg.Brokers...),
		kgo.ClientID(cfg.ClientID),
		kgo.RequiredAcks(kgo.AllISRAcks()),
		kgo.ProducerLinger(0),
	)
	if err != nil {
		return nil, fmt.Errorf("eventbus: new producer: %w", err)
	}
	return &Publisher{client: cl, source: source, log: log}, nil
}

// Publish wraps payload in a canonical Envelope and produces it synchronously to the name's domain
// topic, waiting for the broker ack (at-least-once). The Kafka message key is the entity key, so all
// events for one entity stay ordered on a partition.
func (p *Publisher) Publish(ctx context.Context, name, key string, payload any) error {
	env, err := NewEnvelope(name, key, correlationFrom(ctx), p.source, payload)
	if err != nil {
		return fmt.Errorf("eventbus: build envelope %q: %w", name, err)
	}
	value, err := env.Marshal()
	if err != nil {
		return fmt.Errorf("eventbus: marshal envelope %q: %w", name, err)
	}
	// Start a producer span and stamp the trace context into the record headers (no-op when
	// telemetry is off), so a consumer continues the same trace across the async hop.
	ctx, span := tracer().Start(ctx, "publish "+name, trace.WithSpanKind(trace.SpanKindProducer))
	defer span.End()
	rec := &kgo.Record{Topic: TopicFor(name), Key: []byte(key), Value: value, Headers: injectTraceHeaders(ctx)}
	if err := p.client.ProduceSync(ctx, rec).FirstErr(); err != nil {
		span.RecordError(err)
		return fmt.Errorf("eventbus: produce %q: %w", name, err)
	}
	return nil
}

// Ping checks broker connectivity (for readiness probes).
func (p *Publisher) Ping(ctx context.Context) error { return p.client.Ping(ctx) }

// Close flushes and closes the producer.
func (p *Publisher) Close() { p.client.Close() }

File: linkmint-backend/eventbus-python/src/linkmint_eventbus/publisher.py
"""Async Kafka publisher (aiokafka), byte-compatible with eventbus-go's Publisher."""

from __future__ import annotations

from typing import Any

from aiokafka import AIOKafkaProducer
from opentelemetry import trace
from opentelemetry.propagate import inject
from opentelemetry.trace import SpanKind

from .envelope import Envelope
from .topics import topic_for


class KafkaPublisher:
    """Wraps a payload in a canonical Envelope and produces it synchronously — ``send_and_wait``
    blocks for the broker ack, giving at-least-once semantics."""

    def __init__(self, brokers: list[str], source: str, client_id: str = "linkmint") -> None:
        self._producer = AIOKafkaProducer(
            bootstrap_servers=brokers,
            client_id=client_id,
            acks="all",
            enable_idempotence=True,
        )
        self._source = source
        self._started = False

    async def start(self) -> None:
        await self._producer.start()
        self._started = True

    async def stop(self) -> None:
        if self._started:
            await self._producer.stop()
            self._started = False

    async def publish(
        self,
        name: str,
        key: str,
        payload: dict[str, Any] | None,
        correlation_id: str = "",
    ) -> None:
        env = Envelope.new(
            name=name,
            key=key,
            correlation_id=correlation_id,
            source=self._source,
            payload=payload,
        )
        # Start a producer span and ride the W3C trace context in Kafka record headers (no-op when
        # telemetry is off), so a consumer in any language continues the same trace. The Envelope
        # wire format is untouched — trace context lives only in the transport headers.
        with trace.get_tracer("eventbus").start_as_current_span(
            f"publish {name}", kind=SpanKind.PRODUCER
        ):
            carrier: dict[str, str] = {}
            inject(carrier)
            headers = [(k, v.encode("utf-8")) for k, v in carrier.items()] or None
            await self._producer.send_and_wait(
                topic_for(name),
                value=env.to_canonical_bytes(),
                key=key.encode("utf-8") if key else None,
                headers=headers,
            )
